iqua_softener/iqua.py

```diff
@@ -114,8 +122,34 @@ def get_data(self) -> IquaSoftenerData:
                 salt_level_percent=int(data["saltLevelTenths"]["percent"]),
                 out_of_salt_estimated_days=int(data["outOfSaltEstDays"]["value"]),
                 hardness_grains=int(data["hardnessGrains"]["value"]),
+                water_shutoff_valve_state=water_shutoff_valve_state
             )
 
+    def valve_shutoff(self, state):
+        with requests.Session() as session:
+            if self._token is None or (
+                self._token_expiration_timestamp is not None
+                and datetime.now() > self._token_expiration_timestamp
+            ):
+                self._update_token(session)
+            url = self._get_url(f"system/{self._device_serial_number}/properties")
+            headers = self._get_headers()
+            if state in (0, 1):
+                json = {"waterShutoffValve": state}
+            else:
+                raise IquaSoftenerException(
+                    "Invalid state for water shut off valve."
+                )
+            response = session.put(url, json=json, headers=headers)
+            if response.status_code == 401:
+                self._update_token(session)
+                response = session.put(url, json=json, headers=headers)
+            elif response.status_code != 200:
```

Comment:
I think this should be just `if`.

You want to re-check the status code if you had to re-post the request. Right now, the second request's status code is being ignored.

This check is unnecessarily duplicated in `get_data`.